keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,516 @@
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: restart-replicated-job-example
+  annotations:
+    alpha.jobset.sigs.k8s.io/exclusive-topology: {{topologyDomain}} # 1:1 job replica to topology domain assignment
+spec:
+  # Failure Policy to restart the child jobs of the target ReplicatedJob (gpu-workers) if any fail, but fall
+  # back to the default behavior of restarting the entire JobSet if the driver fails.
+  failurePolicy:
+    rules:
+    - action: RestartReplicatedJob
+      targetReplicatedJobs:
+      - gpu-workers
+    maxRestarts: 10
+  replicatedJobs:
+  - name: driver
+    replicas: 1
+    template:
+      spec:
+        parallelism: 1
+        completions: 1
+        backoffLimit: 0
+        template:
+          spec:
+            restartPolicy: Never
+            containers:
+            - name: main
+              image: python:3.10
+              command: ["..."]
+  - name: gpu-workers
+    replicas: 4 # number of node pools
+    template:
+      spec:
+        parallelism: 2
+        completions: 2
+        backoffLimit: 0
+        template:
+          spec:
+            containers:
+            - name: main
+              image: pytorch:latest
+              command: ["..."]
+            resources:
+              limits:
+                nvidia.com/gpu: 1
+```
+
+### Story 4: FailJob and RestartJob
+
+Dependency: https://github.com/kubernetes/kubernetes/issues/122972
+
+As a user, I want to run a HPC simulation in which each child job runs a simulation with different random initial
+parameters. When a simulation ends, the application will exit with one of two exit codes:
+
+- Exit code 2, which indicates the simulation produced an invalid result due to bad starting parameters, and should
+not be retried.
+- Exit code 3, which indicates the simulation produced an invalid result but the intial parameters were reasonable,
+so the simulation should be restarted.
+
+When a Job fails due to a pod failing with exit code 2, I want the Job to stay in a failed state.
+When a Job fails due to a pod failing with exit code 3, I want to restart the Job.
+
+**Example Failure Policy configuration for this use case**:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: restart-replicated-job-example
+  annotations:
+    alpha.jobset.sigs.k8s.io/exclusive-topology: {{topologyDomain}} # 1:1 job replica to topology domain assignment
+spec:
+  failurePolicy:
+    rules:
+    # If Job fails due to a pod failing with exit code 2, leave it in a failed state.
+    - action: FailJob
+      targetReplicatedJobs:
+      - simulations
+      onJobFailureReasons:
+      - PodFailurePolicy-ExitCode2
+    # If Job fails due to a pod failing with exit code 3, restart that Job.
+    - action: RestartJob
+      targetReplicatedJobs:
+      - simulations
+      onJobFailureReasons:
+      - PodFailurePolicy-ExitCode3
+    maxRestarts: 10
```

Comment:
Good point, updated the example in this story to have 1 podFailurePolicy rule and corresponding JobSet failure policy rule, to `RestartJob` for exit code 3, and a catch all failure policy rule to leave the Job failed otherwise.